include/graaflib/algorithm/strongly_connected_components/kosaraju.tpp

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <graaflib/algorithm/graph_traversal/depth_first_search.h>
+#include <graaflib/algorithm/strongly_connected_components/kosaraju.h>
+#include <graaflib/algorithm/utils.h>
+
+#include <iostream>
+#include <stack>
+
+namespace graaf::algorithm {
+
+template <typename V, typename E>
+std::vector<std::vector<vertex_id_t>> kosarajus_strongly_connected_components(
+    const directed_graph<V, E>& graph) {
+  std::vector<std::vector<vertex_id_t>> sccs;
+
+  if (graph.get_vertices().size() == 0) {
+    return sccs;
+  }
+
+  std::stack<vertex_id_t> stack;
+  std::unordered_set<vertex_id_t> seen_vertices;
+
+  std::function<void(vertex_id_t)> visit_vertex;
+  visit_vertex = [&](vertex_id_t vertex) {
+    if (!seen_vertices.contains(vertex)) {
+      stack.push(vertex);
+      seen_vertices.insert(vertex);
+
+      for (const auto neighbour : graph.get_neighbors(vertex)) {
+        visit_vertex(neighbour);
+      }
+    }
+  };
```

Comment:
In [Kosaraju](https://en.wikipedia.org/wiki/Kosaraju%27s_algorithm), should we not add (prepend) the current vertex to the stack (L in the wikipedia article) **after** visiting its neighbors?

Reply:
I suspect the algorithm might fail for the following graph:
```c++
  directed_graph<int, int> graph{};
  const auto d = graph.add_vertex(4);
  const auto a = graph.add_vertex(1);
  const auto b = graph.add_vertex(2);
  const auto c = graph.add_vertex(3);


  graph.add_edge(a, b, 1);
  graph.add_edge(b, c, 1);
  graph.add_edge(c, a, 1);
  graph.add_edge(b, d, 1);
```

Where we would expect SCCs {A, B, C} and {D}. Might be a nice test to add.

Reply:
I think that's right, that's my bad - will make this tweak!

Reply:
No worries, thanks a lot!